aten/src/ATen/native/RNN.cpp

```diff
@@ -1445,55 +1446,58 @@ std::tuple<Tensor, Tensor, Tensor> lstm(
       TensorList _params, bool has_biases,
       int64_t num_layers, double dropout_p, bool train, bool bidirectional, bool batch_first) {
   TORCH_CHECK(hx.size() == 2, "lstm expects two hidden states");
-  if (at::cudnn_is_acceptable(_input)) {
-    Tensor output, hy, cy;
-    lstm_cudnn_stub(_input.device().type(), output, hy, cy, _input, hx, _params, has_biases,
-            num_layers, dropout_p, train, bidirectional, batch_first);
-    return std::make_tuple(std::move(output), std::move(hy), std::move(cy));
-  }
-#ifdef USE_MPS
-  if (_input.is_mps() && (mps::is_macos_13_or_newer() || num_layers == 1)) {
-    std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor, Tensor> output = at::_lstm_mps(_input, hx, _params, has_biases,
-            num_layers, dropout_p, train, bidirectional, batch_first);
-    std::tuple<Tensor, Tensor, Tensor> return_values = std::make_tuple(std::get<0>(output), std::get<1>(output), std::get<2>(output));
-    return return_values;
-  } else if (_input.is_mps()) {
-    TORCH_WARN_ONCE("Native multi-layer LSTM support in MPS available only on MacOS 13 onwards.",
-                    " Falling back to LSTMCell iteration.",
-                    " This may have performance implications.");
-  }
-#endif
   // if cells are of different size, that means projections are used
   bool has_projections = (hx[0].sym_size(2) != hx[1].sym_size(2));
-  if (use_miopen(_input, dropout_p)) {
-    if (!has_projections) {
+  // Opt for slow path if we are dealing with subclassed tensor.
+  bool any_subclass_like = at::isTensorSubclassLike(_input) || at::areAnyTensorSubclassLike(hx) || at::areAnyTensorSubclassLike(_params);
```

Comment:
Btw, is this regressing how RNNs trace out via make_fx and/or torch.compile?

Reply:
Having a look, will get back in sometime, thanks

Reply:
I think this may affect tensor subclasses which do support access to `data_ptr` as they will take the slower path now.

So we should probably register a decomposition for the `lstm` ops. 
https://github.com/pytorch/pytorch/blob/3a284dae30e1de15d18372c6b448ac67c13d9040/aten/src/ATen/native/native_functions.yaml#L7594-L7599

I think there are 2 options.

**Option 1 : Register in C++**
To choose the composite compliant path for `lstm` operators, we may have to move a few helpers to a header file and then use that for adding the decomposition in C++.

Ref to the required helper template (and there are few other class templates like `FullLayer` and `FullBidirectionalLayer` which we will need to move)
https://github.com/pytorch/pytorch/blob/3a284dae30e1de15d18372c6b448ac67c13d9040/aten/src/ATen/native/RNN.cpp#L1147-L1175 

-----------

**Option 2 : Leverage Python decomposition**
There is already a decomposition in Python but we don't have a mechanism to register for `FuncTorchBatchedDecomposition` key.  (There is a mechanism to register to `FunctorchBatched` key)

https://github.com/pytorch/pytorch/blob/3a284dae30e1de15d18372c6b448ac67c13d9040/torch/_decomp/decompositions.py#L3004-L3008

https://github.com/pytorch/pytorch/blob/3a284dae30e1de15d18372c6b448ac67c13d9040/torch/_functorch/vmap.py#L233-L246

I think Option 2 would be easier. Wdyt? 

Thanks!

Reply:
Oh lol, we should just register the python decomposition to one of the FuncTorchBatched keys

Reply:
IIRC FuncTorchBatchedDecomposition is only for CompositeImplicitAutograd ops, so we'll want to register the decomposition to FuncTorchBatched

Reply:
These `lstm` ops are marked CompositeImplicitAutograd ops.

However, they are not composite compliant for Cudnn (access data_ptr) paths (and maybe other paths like mkldnn, etc)

Reply:
One way to do it is just decompose LSTM at DynamicLayerFrontMode. This would decompose it for all functorch transforms. We'd get decreased performance for functorch.grad, but that's not the worst thing